### PR TITLE
Adjust get neighbor plan

### PR DESCRIPTION
**Pull request by liuyu85cn**
#### What type of PR is this?
- [ ] bug
- [ X] feature
- [ ] enhancement

#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/storage/exec/GetNeighborsNode.h [hunk 1]
```diff
@@ -82,14 +82,24 @@ class GetNeighborsNode : public QueryNode<VertexID> {
       row[1].setList(agg->mutableResult().moveList());
     }
 
-    resultDataSet_->rows.emplace_back(std::move(row));
+    if (!context_->filterInvalidResultOut || context_->resultStat_ == ResultStatus::NORMAL) {
```
Comment (critical27) [inline src/storage/exec/GetNeighborsNode.h:85]: I am not pretty sure about whether we need to check `resultStat_` here. IIRC, we should not check it. You could check it with two edges, one can pass, another can't.

File: src/storage/exec/MultiTagNode.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        return ret;
+      }
+    }
+
+    return nebula::cpp2::ErrorCode::SUCCEEDED;
+  }
+
+  bool valid() const override {
+    // return valid_;
+    auto ret = tagNodes_.back()->valid();
```
Comment (critical27) [inline src/storage/exec/MultiTagNode.h:89]: Is it possible that both tagNodes/edgeNode are empty? If so, perhaps need to check it in processor

**Review by critical27 — COMMENTED** [review]
Generally LGTM~

**Review by wenhaocs — APPROVED** [review]
Just some minor issues. Just want to make sure the logic is "if not edge specified, return only the vertex properties". Can you also give an update of the current behavior of `GetNeighbors` in the description of this PR and `GetNeighborsProcessor.cpp `?

File: src/storage/exec/FilterNode.h [hunk 3]
```diff
@@ -59,27 +64,46 @@ class FilterNode : public IterateNode<T> {
...
+
+  bool checkTagOnly() {
+    auto result = filterExp_->eval(*expCtx_);
+    // NULL is always false
+    auto ret = result.toBool();
+    return ret.isBool() && ret.getBool();
+  }
+
+  // return true when the value iter points to a value which can filter
+  bool checkTagOrEdge() {
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/exec/FilterNode.h]: changed

File: src/storage/exec/GetNeighborsNode.h (hunk 1 above)
Comment (liuyu85cn) [reply src/storage/exec/GetNeighborsNode.h:87]: talked offline.

File: src/storage/test/GetNeighborsTest.cpp [hunk 4]
```diff
@@ -165,7 +187,9 @@ TEST(GetNeighborsTest, PropertyTest) {
 
     ASSERT_EQ(0, (*resp.result_ref()).failed_parts.size());
     // vId, stat, player, serve, teammate, expr
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/test/GetNeighborsTest.cpp:189]: good catch.

File: src/storage/test/GetNeighborsTest.cpp [hunk 5]
```diff
@@ -184,7 +208,9 @@ TEST(GetNeighborsTest, PropertyTest) {
 
     ASSERT_EQ(0, (*resp.result_ref()).failed_parts.size());
     // vId, stat, player, - teammate, - serve, expr
-    QueryTestUtils::checkResponse(*resp.vertices_ref(), vertices, over, tags, edges, 1, 6);
+    size_t expectColumnCount = 4;
```
Comment (critical27) [inline src/storage/test/GetNeighborsTest.cpp:211]: update the comment as well, so as `expectColumnCount` you add in this PR

File: src/storage/test/GetNeighborsTest.cpp [hunk 6]
```diff
@@ -1014,20 +1066,22 @@ TEST(GetNeighborsTest, TtlTest) {
...
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[2].empty());   // player expired
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[3].empty());   // team not exists
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[4].empty());   // general tag not exists
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[5].isList());  // - teammate valid
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[6].empty());   // - serve expired
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[7].empty());   // + serve expired
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[8].isList());  // + teammate valid
-    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[9].empty());   // expr
+    ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[2].empty());   // expr player expired
+    // ASSERT_TRUE((*resp.vertices_ref()).rows[0].values[3].empty());   // team not exists
```
Comment (critical27) [inline src/storage/test/GetNeighborsTest.cpp:1078]: Anything special in this case?

File: src/storage/test/GetNeighborsTest.cpp [hunk 7]
```diff
@@ -49,6 +49,28 @@ TEST(GetNeighborsTest, PropertyTest) {
...
+    bool retNoneProps = false;
+    auto req = QueryTestUtils::buildRequest(totalParts, vertices, over, tags, edges, retNoneProps);
+
+    auto* processor = GetNeighborsProcessor::instance(env, nullptr, threadPool.get());
+    auto fut = processor->getFuture();
+    processor->process(req);
+    auto resp = std::move(fut).get();
+
+    ASSERT_EQ(0, (*resp.result_ref()).failed_parts.size());
+    // vId, stat, player, serve, expr
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/test/GetNeighborsTest.cpp]: fixed.

File: src/storage/CommonUtils.h [hunk 8]
```diff
@@ -223,6 +223,10 @@ struct RuntimeContext {
   // used for update
   bool insert_ = false;
 
+  // some times, one line is filter out but still return (has edge)
+  // and some time, this line is just removed from thre return result
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/CommonUtils.h:227]: thx.

File: src/storage/test/GetNeighborsTest.cpp (hunk 5 above)
Comment (liuyu85cn) [reply src/storage/test/GetNeighborsTest.cpp:211]: ok

**Review by critical27 — APPROVED** [review]
Well done~

**Review by panda-sheep — DISMISSED** [review]
Good!

File: src/storage/query/GetNeighborsProcessor.cpp [hunk 9]
```diff
@@ -184,24 +186,32 @@ StoragePlan<VertexID> GetNeighborsProcessor::buildPlan(RuntimeContext* context,
...
+|     TagNodes     |        |     EdgeNodes    |               +------+-----+
++------------------+        +------------------+                      |
+                                                               +------+-----+
+                                                               |MultiTagNode|
+                                                               +------+-----+
+                                                                      |
+                                                                 +----+---+
+                                                                 |TagNodes|
+                                                                 +--------+
+
```
Comment (panda-sheep) [inline src/storage/query/GetNeighborsProcessor.cpp:214]: 👍

**Review by panda-sheep — APPROVED** [review]

**Review by critical27 — APPROVED** [review]

**Pull request merged**